rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml

```diff
@@ -0,0 +1,51 @@
+apiVersion: monitoring.coreos.com/v1
+kind: PrometheusRule
+metadata:
+  name: rhtap-pipeline-alerting
+  labels:
+    tenant: rhtap
+spec:
+  groups:
+    - name: pipeline_alerts
+      interval: 1m
+      rules:
+        - alert: HighSchedulingOverhead
+          expr: |
```

Comment:
We're summarizing across clusters. Is this the intention?

Reply:
I think we should use either `increase` 
 (https://prometheus.io/docs/prometheus/latest/querying/functions/#increase) or maybe `histogram_count` and `histogram_sum` (https://prometheus.io/docs/prometheus/latest/querying/functions/#histogram_count-and-histogram_sum). I Never tried the latter.

If I understand correctly, we currently assume that the metrics are monotonically growing, but that might not always be the case.
The use of, at least, `increase` should help overcoming counter resets etc.

Reply:
I think we should use either `increase` 
 (https://prometheus.io/docs/prometheus/latest/querying/functions/#increase) or maybe `histogram_count` and `histogram_sum` (https://prometheus.io/docs/prometheus/latest/querying/functions/#histogram_count-and-histogram_sum). I Never tried the latter.

If I understand correctly, we currently assume that the metrics are monotonically growing, but that might not always be the case.
The use of, at least, `increase` should help overcoming counter resets etc.

(to be honest, I can't say I fully understand the metrics yet, but I tend to think that `sum_over_time` might not be the way to go here, as `_sum` and `_count` should behave as counters -- as far as I understand)

Reply:
Unintentional.

I do see `{{ $labels.source_cluster }}` in the descriptions of some of the existing alerts.  Is that how I make it "per cluster" @yftacherzog ?

Reply:
On the second point, 
- simply using sum was not sufficient through manual testing with the OCP metric console with these histograms.  The combination was needed.
- that said, yes, my observation over many weeks has been that the metric only grows until the controller I'm using to measure is restarted.  If that is what you mean by a reset, at least with the 30 minute scrape window, that had been a non issue.
- I will research the increase / histogram_count / histogram_sum functions and if what they produces reconciles with me, I'll switch to those.  If they do not reconcile, I'll describe what is wrong, and we can continue from there.

Reply:
> Unintentional.
> 
> I do see `{{ $labels.source_cluster }}` in the descriptions of some of the existing alerts. Is that how I make it "per cluster" @yftacherzog ?

`{{ $labels.source_cluster }}` is the label that tells us the where the metric came from. We should have it in the summary/description, but more importantly, the query we did should make the calculations per-cluster (assuming that is what we're after). It means that if we use `sum` then we need to make sure time series with different source_cluster are not summed together.

Reply:
> On the second point,
> 
> * simply using sum was not sufficient through manual testing with the OCP metric console with these histograms.  The combination was needed.
> * that said, yes, my observation over many weeks has been that the metric only grows until the controller I'm using to measure is restarted.  If that is what you mean by a reset, at least with the 30 minute scrape window, that had been a non issue.
> * I will research the increase / histogram_count / histogram_sum functions and if what they produces reconciles with me, I'll switch to those.  If they do not reconcile, I'll describe what is wrong, and we can continue from there.

Yes. I meant counter reset due to something like controller restart. There are certain query functions Prometheus offers (e.g. `increase`) that should be able to detect counter resets and give a correct value even when the counter resets.

Reply:
So I logged onto the RHTAP Prod RH cluster at https://console-openshift-console.apps.stone-prd-rh01.pg1f.p1.openshiftapps.com and I went to Observer-> Metrics

1) neither `histogram_count` nor `histogram_sum` were available as functions; my first assumption there is that those clusters are not a a sufficient level wrt prometheus to container those features.  I'm not sure how the flow from the App SRE grafana to the individual clusters goes, and where those 2 functions are used, but it seem like if it does not work on the local cluster, those are not available to us.
2) `increase` does exist.... after analyzing what `increase(...)` produces I was able to modify one of our overhead queries where I replaced `sum_over_time` with increase, and get an scheduling overhead ratio that was believable

```
sum(increase(pipelinerun_duration_scheduled_seconds_sum{status='succeded'}[30m])) / sum(increase(pipelinerun_duration_scheduled_seconds_count{status='succeded'}[30m])) / sum(increase(tekton_pipelines_controller_pipelinerun_duration_seconds_sum{status='success'}[30m])) / sum(increase(tekton_pipelines_controller_pipelinerun_duration_seconds_count{status='success'}[30m])) * 100
```

So minimally, I'm fine with replacing `sum_over_time` with `increase`.  Now, of course, you see I still need to use `sum` in order to sufficiently aggregate the various _sum and _count observations of the histograms we are using in order to calculate an overall percentage.

However, wrt using `sum` and making sure different source clusters are not summed together, I can only make educated guesses as to how to filter.  Do I add a `cluster` label in the query?  Or is there something else?  Also, in looking at the other existing alerts, I see no such cluster filtering, and am unclear as to why the use of `sum` makes any difference wrt using data from multiple clusters?.  Aren't those other queries technically being run against multiple clusters?

Ultimately the alerts are defined on the individual clusters, no?  Or is this a question of the App SRE grafana dashbaord running outside our clusters, and being able to query against multilpe clusters, and hence we want what the app sre dashboard shows to match what the individual cluster is alerting on?

Lastly, eventually, RHTAP is suppose to go down to one production cluster, so it might ultimately be a moot point.  But ideally we sort out the above in case that ever changes.

Thanks

Reply:
Or should I be using something other than `sum` to take all the entries in the vectors produced by `increase` into single values ?

Reply:
The alerts are managed in RHOBS which contains metrics coming from all data plane (and control plane clusters).

When you aggregate (e.g. with `sum`), you can specify according to which labels you're going to aggregate. you do that using the `by` and `without` keywords. It will make Prometheus keep only or everything-but the labels you specify. I think (always get confused with that) you'd need to use something like:
```
sum by (source_cluster) (my_metric{my_label="foo", another_label="bar"...})
```
in this case. Give it a try and see what you get back.

See this reference for more details: https://prometheus.io/docs/prometheus/latest/querying/operators/#aggregation-operators

About whether or not we should sum the clusters together, I leave it for you to decide.

Reply:
OK thanks for the pointer.  Back over at https://console-openshift-console.apps.stone-stg-m01.7ayg.p1.openshiftapps.com and the metrics browser with its auto-completion I discovered that `sum by (cluster_id)` is a thing and it works correctly when I add it to my queries as is (i.e. still `sum_over_time` vs. `increase`)

`make check-and-test ` also passed when I added it to the alerts.

Now, given the `source_cluster` label already noted above, I'm guessing that when done from RHOBS and not a single cluster's metrics browser, I should use `source_cluster` as you noted vs. `cluster_id`.  `make check-and-test` passed with that as well.

Regardless of what I did with `sum...` when I switched my unit tests from `sum_over_time` to `increase` the `make check-and-test` started to fail.  I'll see if I can sort that out as I manipulate the time series in the tests per the thread below.